File: raceTimeNode/services/analyticsService.js
/**
 * Sliding Window Algorithm
 * This function finds how many groups of 'k' laps add up to 'x'.
 * * @param {Array} laps - An array of integers (e.g., [1000, 2000, 3000])
 * @param {Number} k - The size of the window (number of contiguous laps)
 * @param {Number} x - The target sum we are looking for
 */
const countWorkoutPatterns = (laps, k, x) => {
  // 1. Initialize variables
  let left = 0;         // The trailing edge of our window
  let currentSum = 0;   // The running total of the current window
  let count = 0;        // The number of matches found

  // 2. Loop through the array using 'right' as the leading edge
  for (let right = 0; right < laps.length; right++) {
    
    // 3. Add the newest lap on the right to our running total
    currentSum += laps[right];

    /**
     * 4. SHRINK THE WINDOW
     * If our window has grown larger than size 'k', we must 
     * remove the element at the 'left' index to keep the window at size 'k'.
     */
    if (right >= k) {
      currentSum -= laps[left]; // Subtract the oldest lap
      left++;                    // Move the left boundary forward
    }

    /**
     * 5. CHECK THE TARGET
     * Once our window has reached the full size of 'k' (at index k-1),
     * we check if the sum matches our target value 'x'.
     */
    if (right >= k - 1) {
      if (currentSum === x) {
        count++; // We found a matching pattern!
      }
    }
  }

  // 6. Return the final result to the controller
  return count;
};

/**
 * analyticsService.js
 * Logic to find windows with an even sum.
 */
const getEvenSumWindows = (laps, k) => {
  // 1. Transform the array of lap objects into an array of just numbers
  // This is like the .stream().mapToInt() in Java
//   const arr = laps.map(lap => lap.durationMs);

const arr = laps.map(lap => {
    const rawLap = lap.get ? lap.get({ plain: true }) : lap;
    return rawLap.durationMs;
  });
  
  console.log("Raw Numbers for Math:", arr); // Check if this is [1000, 1200...] or [undefined, undefined...]

  const results = []; // This will hold our "WindowResult" objects
  let currentSum = 0;

  // 2. Sliding Window Loop
  for (let right = 0; right < arr.length; right++) {
    currentSum += arr[right];

    // 3. If window size exceeds K, subtract the leftmost element
    if (right >= k) {
      currentSum -= arr[right - k];
    }

    // 4. Once we have a full window of size K
    if (right >= k - 1) {
      // Check if the sum is even using Modulo
      if (currentSum % 2 === 0) {
        // 5. Capture the window elements
        // .slice(start, end) creates the "snapshot" like Java's subList
        const windowElements = arr.slice(right - k + 1, right + 1);

        // 6. Push a new object into the results array
        // (This replaces the 'new WindowResult()' part of Java)
        results.push({
          elements: windowElements,
          totalSum: currentSum
        });
      }
    }
  }

  return results;
};

// module.exports = { getEvenSumWindows };

// Export the function so it can be imported by the LapController
module.exports = {
 // countWorkoutPatterns,
  getEvenSumWindows
};

File: raceTime/raceTime/src/main/java/com/micro/raceTime/service/AnalyticsService.java
package com.micro.raceTime.service; // Defines the location of this file in the project

// --- IMPORTS: Bringing in the tools we need ---
import com.micro.raceTime.model.Lap; // The object representing a single race lap
import com.micro.raceTime.model.WindowResult; // The object we use to bundle results
import org.springframework.beans.factory.annotation.Autowired; // For dependency injection
import org.springframework.kafka.core.KafkaTemplate; // The tool used to send messages to Kafka
import org.springframework.kafka.support.SendResult; // Holds the result of a Kafka send attempt
import org.springframework.stereotype.Service; // Marks this as a Spring-managed service class

import java.util.ArrayList; // Used to create dynamic lists
import java.util.List; // The standard interface for lists
import java.util.concurrent.CompletableFuture; // For handling background (asynchronous) tasks
import java.util.stream.Collectors; // Used to transform data collections

@Service // Tells Spring to create and manage one instance of this class
public class AnalyticsService {

    @Autowired // Automatically connects the Kafka tool from the Spring context
    private KafkaTemplate<String, WindowResult> kafkaTemplate; // Our "messenger" to send data to Kafka

    /**
     * Main method: Takes a list of laps and a window size 'k', finds even sums, and alerts Kafka.
     */
    public List<WindowResult> getEvenSumWindows(List<Lap> laps, int k) {

        // Transform the list of Lap objects into a simple list of numbers (millisecond durations)
        List<Integer> arr = laps.stream()
                .map(Lap::getDurationMs) // Reach inside each lap and grab the time
                .collect(Collectors.toList()); // Put all those times into a new list

        List<WindowResult> results = new ArrayList<>(); // A list to store all the even windows we find
        int currentSum = 0; // A variable to keep track of the total sum inside our "window"

        // Loop through the array from left to right (this is our 'sliding window')
        for (int right = 0; right < arr.size(); right++) {

            currentSum += arr.get(right); // Add the newest element on the right to our sum

            // If the window has grown larger than size 'k', we must remove the oldest element
            if (right >= k) {
                currentSum -= arr.get(right - k); // Subtract the element that just fell out of the window
            }

            // Once we have processed at least 'k' elements, our window is full and ready to check
            if (right >= k - 1) {

                // Check if the current window total is an even number (remainder of division by 2 is 0)
                if (currentSum % 2 == 0) {

                    // Create a sub-list containing exactly the elements currently inside the window
                    List<Integer> windowElements = new ArrayList<>(arr.subList(right - k + 1, right + 1));

                    // Package the elements and their sum into a Result object
                    WindowResult foundWindow = new WindowResult(windowElements, currentSum);

                    results.add(foundWindow); // Add this finding to our local list to return to the caller

                    // Call our helper method to broadcast this finding to the rest of the system via Kafka
                    sendGoalToKafka(foundWindow);
                }
            }
        }
        return results; // Return the full list of even windows found during the process
    }

    /**
     * Helper Method: Handles the technical details of talking to the Kafka Broker.
     */
    private void sendGoalToKafka(WindowResult result) {

        // Start the send process to the "goal-topic". This happens in the background.
        CompletableFuture<SendResult<String, WindowResult>> future =
                kafkaTemplate.send("goal-topic", result);

        // Wait for Kafka to respond with either a "Success" or an "Error"
        future.whenComplete((success, ex) -> {

            if (ex == null) {
                // If ex (exception) is null, it worked! Print where the message was stored.
                System.out.println("✅ Sent to Kafka! Offset: " + success.getRecordMetadata().offset());
            } else {
                // If there is an exception, the broker might be down. Log the error message.
                System.err.println("❌ Kafka Error: " + ex.getMessage());
            }
        });
    }
}